message_ix/tests/test_feature_vintage_and_active_years.py

```diff
@@ -0,0 +1,203 @@
+
+    if append is not None:
+        result = pd.concat([result, append]).sort_values(
+            ["year_vtg", "year_act"], ignore_index=True
+        )
+
+    return result
+
+
+def test_vintage_and_active_years1(test_mp):
+    """Basic functionality of :meth:`.vintage_and_active_years`."""
+    years = (2000, 2010, 2020)
+    fmy = years[1]
+
+    # Set up scenario, tech, and retrieve valid (yv, ya) pairs
+    scen, yvya_all = _setup(test_mp, years, fmy)
+
+    # Default / no arguments
+    assert_frame_equal(
+        _q(yvya_all, f"year_act >= {fmy}"),
+        scen.vintage_and_active_years(),
+    )
+
+    # part is before horizon
+    obs = scen.vintage_and_active_years(ya_args=("foo", "bar", "2000"))
+    exp = pd.DataFrame({"year_vtg": (2000,), "year_act": (2010,)})
+    assert_frame_equal(exp, obs)
+
+    obs = scen.vintage_and_active_years(
+        ya_args=("foo", "bar", "2000"), in_horizon=False
+    )
+    exp = pd.DataFrame({"year_vtg": (2000, 2000), "year_act": (2000, 2010)})
+    assert_frame_equal(exp, obs)
+
+    # fully in horizon
+    obs = scen.vintage_and_active_years(ya_args=("foo", "bar", "2010"))
+    exp = pd.DataFrame({"year_vtg": (2010, 2010), "year_act": (2010, 2020)})
+    assert_frame_equal(exp, obs)
+
+    # part after horizon
+    obs = scen.vintage_and_active_years(ya_args=("foo", "bar", "2020"))
+    exp = pd.DataFrame({"year_vtg": (2020,), "year_act": (2020,)})
+    assert_frame_equal(exp, obs)
+
+    # Advance the first model year
+    scen.add_cat("year", "firstmodelyear", years[-1], is_unique=True)
+
+    # Empty data frame: only 2000 and 2010 valid year_act for this node/tec; but both
+    # are before the first model year
+    obs = scen.vintage_and_active_years(
+        ya_args=("foo", "bar", years[0]), in_horizon=True
+    )
+    assert_frame_equal(
+        pd.DataFrame(columns=["year_vtg", "year_act"]), obs, check_dtype=False
+    )
+
+    # Exception is raised for incorrect arguments
+    with pytest.raises(
+        ValueError, match="At least 2 arguments are required if using `ya_args`"
+    ):
+        scen.vintage_and_active_years(ya_args=("foo"))
+
+
+def test_vintage_and_active_years2(test_mp):
+    """:meth:`.vintage_and_active_years` with periods of uneven duration."""
+    years = (2000, 2005, 2010, 2015, 2020, 2030)
+    fmy = years[2]
+
+    scen, yvya_all = _setup(test_mp, years, fmy)
+
+    extra = pd.Series(dict(year_vtg=2010, year_act=2030)).to_frame().T
+
+    # Check if default function call is valid
+    obs = scen.vintage_and_active_years()
+    exp = _q(yvya_all, f"year_act >= {fmy}")
+    assert_frame_equal(exp, obs)
+
+    # Check standard functionality with different period duration lengths
+    obs = scen.vintage_and_active_years(ya_args=("foo", "bar", "2010"))
+    exp = _q(yvya_all, f"year_vtg == 2010 and year_act >= {fmy}")
+    assert_frame_equal(exp, obs)
+
+    # Check if no vintge-year is passed, that all values corresponding to technical
```

Comment:
With the preparation of `exp` now slimmed down a lot, we can see that this and the next three assertions each have an identical comment above them. This doesn't help understand what is *different* between them.

Reply:
I rewrote these comments.